content/blog/2023-12-11-arrays-and-variable-arguments-in-Bril.md

```diff
@@ -0,0 +1,96 @@
+
+I also changed the existing assignment case to allow assignment to an element in the array. This is similar to evaluating indexed array as an rvalue, mentioned above.
+
+It is the programmer's responsibility to free the arrays, unlike in real TypeScript, because Bril garbage management is manual by default. I overloaded the existing `free` function in `mem.d.ts` so that the programmer can free the array in TypeScript.
+```
+export function free<T>(array: T): void;
+```
+
+The code below 
+```
+const numbers: bigint[] = [18n,26n];
+console.log(numbers[0n])
+free<bigint[]>(numbers);
+```
+gets compiled to 
+```
+@main {
+  v0: int = const 2;
+  v1: ptr<int> = alloc v0;
+  cur: ptr<int> = id v1;
+  v2: int = const 1;
+  v3: int = const 18;
+  store cur v3;
+  cur: ptr<int> = ptradd cur v2;
+  v4: int = const 26;
+  store cur v4;
+  numbers: ptr<int> = id v1;
+  v5: ptr<int> = id numbers;
+  v6: int = const 0;
+  v7: ptr<int> = ptradd v5 v6;
+  v8: int = load v7;
+  print v8;
+  v9: int = const 0;
+  v10: ptr<int> = id numbers;
+  free v10;
+  v11: int = const 0;
+}
+```
+
+As you can see, there is quite a bit of redundant and dead code due to the way expressions are evaluated in the compiler, but it should be easily eliminated with optimizations.
+
+
+# What were the hardest parts to get right?
+The compiler extension was much trickier than the variable argument one. For the longest time, I did not realize that the link to the declaration file `typescript.d.ts` was broken, because the link to the rest of the code was not. This meant that the compiler was mostly working without it except in a few places, and when I added arrays I thought I must have done something wrong. I wasn't really familiar with the setup of TypeScript and there is very little documentation and few StackOverflow posts on the public TS compiler API, so I was stuck on that part for awhile. It also took some time to find the correct `functions` or `SyntaxKind` for my use cases, as I had to comb through the TypeScript codebase or rely on print statements. For anyone who is going to work with the TypeScript AST, I highly recommend the [AST viewer](https://ts-ast-viewer.com/#).  
```

Comment:
I had not seen that AST viewer! That's pretty cool. It would have saved me a lot of time in my days of hacking on TS compiler internals…